src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1632,54 +1614,48 @@ cpp2::ErrorCode RaftPart::verifyLeader(
-                req.get_leader_ip() == leader_.first &&
-                req.get_leader_port() == leader_.second) {
-                VLOG(3) << idStr_ << "Same leader";
-                return cpp2::ErrorCode::SUCCEEDED;
-            }
-            break;
-        }
-        case Role::LEADER: {
-            // In this case, the remote term has to be newer
-            // TODO optimize the case that the current partition is
-            // isolated and the term keeps going up
-            break;
-        }
-        case Role::CANDIDATE: {
-            // Since the current partition is a candidate, the remote
-            // term has to be newer so that it can be accepted
-            break;
-        }
-    }
-
     // Make sure the remote term is greater than local's
     if (req.get_current_term() < term_) {
-        LOG_EVERY_N(ERROR, 100) << idStr_
-                                << "The current role is " << roleStr(role_)
-                                << ". The local term is " << term_
-                                << ". The remote term is not newer";
+        LOG_EVERY_N(INFO, 100) << idStr_
+                               << "The current role is " << roleStr(role_)
+                               << ". The local term is " << term_
+                               << ". The remote term is not newer";
         return cpp2::ErrorCode::E_TERM_OUT_OF_DATE;
-    }
-    if (role_ == Role::FOLLOWER || role_ == Role::LEARNER) {
-        if (req.get_current_term() == term_ && leader_ != std::make_pair(0, 0)) {
-            LOG(ERROR) << idStr_ << "The local term is same as remote term " << term_
-                       << ". But I believe leader exists.";
-            return cpp2::ErrorCode::E_TERM_OUT_OF_DATE;
+    } else if (req.get_current_term() > term_) {
+        // Leader stickness, no matter the term in Request is larger or not.
+        // TODO(heng) Maybe we should reconsider the logic
+        if (leader_ !=  std::make_pair(0, 0) && leader_ != candidate
+                && lastMsgRecvDur_.elapsedInMSec()
+                        < FLAGS_raft_heartbeat_interval_secs * 1000) {
+            LOG_EVERY_N(INFO, 100) << idStr_ << "I believe the leader " << leader_ << " exists. "
+                                   << "Refuse to append logs of " << candidate;
+            return cpp2::ErrorCode::E_WRONG_LEADER;
         }
+    } else {
+        // req.get_current_term() == term_
+        do {
+            if (role_ != Role::LEADER
+                    && leader_ == std::make_pair(0, 0)) {
+                LOG_EVERY_N(INFO, 100) << idStr_
+                                       << "I dont know who is leader for current term " << term_
+                                       << ", so accept the candidate " << candidate;
+                break;
+            }
+            // Same leader
+            if (role_ != Role::LEADER
+                    && candidate == leader_)  {
+                return cpp2::ErrorCode::SUCCEEDED;
```

Comment:
maybe in most case will return from this branch, is it better do this check first?

Reply:
That's fine.  The branch elimination is not so important in current path.